### PR TITLE
Establish request and response validation outside of middlewares

### PR DESCRIPTION
**Pull request by ahx**
Working towards 1.0.

Request validation
```ruby
OpenapiFirst.load('petstore.yaml').request(rack_request).validate
```

Response validation
```ruby
OpenapiFirst.load('petstore.yaml').request(rack_request).response(rack_response).validate!
```

Configuration:
```ruby
OpenapiFirst.configure do |config|
  config.request_validation_error_response = :jsonapi
end
```

Middlewares:
```ruby
 use OpenapiFirst::RequestValidation::Middleware, 'petstore.yaml'
```

etc...

### CODE REVIEW

File: lib/openapi_first/response_validation.rb [hunk 1]
```diff
@@ -16,98 +23,8 @@ def call(env)
       return @app.call(env) unless operation
 
       response = @app.call(env)
-      validate(response, operation)
+      Validator.new(operation).validate(response)
```
Comment (MrBananaLord) [inline lib/openapi_first/response_validation.rb:26]: I like that! I suppose (since this is a draft) you are going to have similar approach for the request validation? :)

File: lib/openapi_first/request_validation/request_body_validator.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+require_relative 'body_parser'
+
+module OpenapiFirst
+  module RequestValidation
+    class RequestBodyValidator
+      def initialize(operation, env)
+        @operation = operation
+        @env = env
```
Comment (MrBananaLord) [inline lib/openapi_first/request_validation/request_body_validator.rb:10]: what do you think about passing only required data instead of the whole env? e.g.
```ruby
def initialize(operation, body, content_type = "application/json")
```

passing `env` around seems... like a little bit too much information passed when compared to the actual data being used. Also, the interface is less explicit

File: spec/definition/operation_spec.rb [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    describe 'when status code cannot be found' do
+      let(:spec) { OpenapiFirst.load('./spec/data/parameters.yaml') }
+      let(:operation) { spec.operations.last }
+
+      it 'returns nil' do
+        expect(operation.response_for(201, 'application/json')).to be_nil
+      end
+    end
+
+    describe 'when response object media type cannot be found' do
```
[the remark this answers is not in the document]
Comment (ahx) [reply spec/definition/operation_spec.rb:127]: Good point. Will change to context.

File: lib/openapi_first/request_validation/request_body_validator.rb (hunk 2 above)
Comment (ahx) [reply lib/openapi_first/request_validation/request_body_validator.rb:10]: Sounds good!
But be aware that RequestBodyValidator is not part of the "public" API. I would like to align request and response validation a little and remove some Openapi 3 jargon like "operation" to this one will probably resolve in something like a body validator or content validation ([contentType + contentSchema](https://github.com/OAI/moonwalk#requests))

File: lib/openapi_first/response_validation.rb (hunk 1 above)
Comment (ahx) [reply lib/openapi_first/response_validation.rb]: Yes. Very similar. :)

**Comment by ahx** [comment]
I think this is ready to go. Can you give it a try? Maybe @MrBananaLord or @supermomonga?

File: lib/openapi_first/request_validation/request_body_validator.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        request_body = operation.request_body
+        return unless request_body
+
+        schema = request_body.schema_for(request_content_type)
+        unless schema
+          RequestValidation.fail!(:unsupported_media_type,
+                                  message: "#{request_content_type.inspect} is not defined.")
+        end
+
+        RequestValidation.fail!(:invalid_body) if request_body.required? && parsed_request_body.nil?
```
Comment (MrBananaLord) [inline lib/openapi_first/request_validation/request_body_validator.rb:20]: this results in an error with the message "Request body invalid: ". I'd go with something more descriptive such as:
```suggestion
        RequestValidation.fail!(:invalid_body, message: "request body is not defined.") if request_body.required? && parsed_request_body.nil?
```

File: lib/openapi_first/request_validation/request_body_validator.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+module OpenapiFirst
+  module RequestValidation
+    class RequestBodyValidator
+      def initialize(operation)
+        @operation = operation
+      end
+
+      def validate!(parsed_request_body, request_content_type)
+        request_body = operation.request_body
```
Comment (MrBananaLord) [inline lib/openapi_first/request_validation/request_body_validator.rb:11]: this is interesting... what if the body is not defined in the OAS but it's sent in the request? 🤔

**Review by MrBananaLord — APPROVED** [review]
Works fine for me! The only minor difference I found is the error message when the request body is expected, but not provided :)

File: lib/openapi_first/request_validation/request_body_validator.rb (hunk 5 above)
Comment (ahx) [reply lib/openapi_first/request_validation/request_body_validator.rb:11]: In that case request validation should pass, because you don't say that request body must be empty. I'll add a test about that in case there is none already.

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,18 +2,28 @@
 
 ## Unreleased
 
+- Fix response header validation with Rack 3
+- Add interface to validate requests / responses without middlewares (see "Manual validation" in README)
+- Breaking: Rename OpenapiFirst::ResponseInvalid to OpenapiFirst::ResponseInvalidError
```
[the remark this answers is not in the document]
Comment (ahx) [reply CHANGELOG.md:7]: Thanks. Good catch. Let's keep OpenapiFirst::Error

**Comment by ahx** [comment]
I still would like to 
- Change the default error response
- Make `ResponseValidation::Validator#validate` (not `#validate!`) also return a Failure, just like in request validation

…but that is a story for another PR.
